homeassistant/components/sonos/speaker.py

```diff
@@ -591,13 +591,20 @@ def speaker_activity(self, source: str) -> None:
             self.async_write_entity_states()
             self.hass.async_create_task(self.async_subscribe())
 
-    async def async_check_activity(self, now: datetime.datetime) -> None:
+    @callback
+    def async_check_activity(self, now: datetime.datetime) -> None:
         """Validate availability of the speaker based on recent activity."""
         if not self.available:
             return
         if time.monotonic() - self._last_activity < AVAILABILITY_TIMEOUT:
             return
+        # Ensure the ping is canceled at shutdown
+        self.hass.async_create_background_task(
```

Comment:
Should we also cancel the ping when unloading the entry?

Reply:
Why does it work on shutdown?

Reply:
But we don't know that we're the last integration that consume a thread in the executor when shutting down, right? Isn't there still a risk that the executor will be exhausted during shutdown if we allow integrations to behave that way?

Reply:
We only force an exception to be raised in the executor shutdown after we have already tried to cleanly join the thread after a shutdown. 

At that point the executor is no longer accepting new jobs we only force the exception to be raise to terminate the lingering threads that are holding up shutdown.

https://github.com/home-assistant/core/blob/dev/homeassistant/util/executor.py

Reply:
Right. But we first try to cancel the background tasks, right? If we allow integrations to use the executor and not be responsible when we try to cancel that task we'll potentially exhaust the executor before reaching the executor shutdown.

Am I following the shutdown flow correctly?

Reply:
I'm working on the premies that executor jobs must run to completion or be forced to raise an exception when we shutdown the executor.

### Current behavior before this change at shutdown when `async_check_activity` is running at shutdown and waiting for the executor job:

Background tasks are canceled here which **does not** include this task
https://github.com/home-assistant/core/blob/c566303edbfd2180d3c068aa776e65d8136eb21b/homeassistant/core.py#L727

We block here waiting for the task and the executor job its awaiting to finish
https://github.com/home-assistant/core/blob/c566303edbfd2180d3c068aa776e65d8136eb21b/homeassistant/core.py#L739

### New behavior before this change at shutdown when `async_check_activity` is running at shutdown and waiting for the executor job:

Background tasks is canceled here which now includes this task
https://github.com/home-assistant/core/blob/c566303edbfd2180d3c068aa776e65d8136eb21b/homeassistant/core.py#L727

We continue on here since the task has been canceled and the cancelation propagates
https://github.com/home-assistant/core/blob/c566303edbfd2180d3c068aa776e65d8136eb21b/homeassistant/core.py#L739

`async_stop` now finishes https://github.com/home-assistant/core/blob/c566303edbfd2180d3c068aa776e65d8136eb21b/homeassistant/core.py#L813

which gets us to 
https://github.com/home-assistant/core/blob/c566303edbfd2180d3c068aa776e65d8136eb21b/homeassistant/runner.py#L184

If the executor thread is still doing the ping it gets killed at this point.

Reply:
There may still be a problem as I'm not sure if `async_block_till_done` still blocks on canceled tasks but its a bit outside the scope of this PR

Reply:
Ok, I understand now that this is a general problem of tasks in threads in the executor.

I don't see a good way of avoiding the problem I was thinking about. But its probably not a big problem.